### ISSUE TITLE
[MU4 Issue] Tab Stops on New Score | Additional score Information

### ISSUE THREAD

**Issue by critchie** [musescore/MuseScore#12851]
**Describe the bug**
When pressing tab from the Title edit box it advances to the Cancel button.

**To Reproduce**
1. Create a new score
2. Work your way to the “Additional score information” page
3. Click in the Title edit box
4. Press Tab button

**Expected behavior**
I expect the cursor to advance to the Composer edit box. A Tab press from any of edit boxes advances to the Cancel button. I expect the cursor to advance to the next field. Pretty standard UI behavior and easily overlooked while more important features are fleshed out.

**Platform information**
 - OS: MacOS

**Comment by MarcSabatella** [comment]
I *believe* this is by design, although it's definitely worth further discussion.  The way MuseScore 4 is organized for accessibility, F6 is the highest level shortcut - takes you from one major area of a window to another.  Tab takes you between *groupings* within a major area.  Cursor keys move you *within* groupings.  So in this case, Down is how to reach Composer.

It's common for any sufficiently complex application window to use a similar hierarchy, so at that level, nothing seems wrong with the current scheme.  And yet, I don't know that every single dialog in MuseScore needs that kind of hierarchy.  Many could work with just Tab, or F6 and Tab.  It will be interesting to get more feedback on this!

**Comment by critchie** [comment]
Not using Tabs for navigation between fields seems to go against years and years of HMI standards. I clicked around and the dialogs under the Format menu item, like Styles, use Tab to move around which is what I’d expect.

I did notice on MacOS you do have to opt-in to full keyboard accessibility to get Tab to move between all interface widgets.

**Comment by MarcSabatella** [comment]
Yes, in the past that is how programs normally worked, but times are change.  See for instance this - https://musescore.org/en/node/305103

(Joanie Diggs is one of the developers of the screen reader Orca and can be considered an expert on accessibility)

I'm not saying there aren't tradeoffs to consider, but, it is important to note that this is not something that was done for no reason or without precedent.

**Comment by critchie** [comment]
I’m not sure that article applies in this case. When it comes to forms with edit boxes, check boxes, radio boxes, drop downs and other input components the Tab button is the standard for navigating between them. The MacOS and Windows HMI guidelines spell this out clearly. I think changing this behavior is right up there with changing, say, Ctrl+C for copy.

With menus, arrow keys are standard. The standard for getting to the application menu is Ctrl+F2 (MacOS)/F10 (Windows) and once the menu has focus the arrow keys can be used to select the menu like File, Edit, View, etc.

**Comment by MarcSabatella** [comment]
Like I said, definitely room for discussion.  I was surprised by this too at first.

**Comment by MarcSabatella** [comment]
I’m not sure which Windows document you mean, but this one pretty clearly sanctions use of Tab to move between panes:

https://docs.microsoft.com/en-us/previous-versions/windows/desktop/dnacc/guidelines-for-keyboard-user-interface-design

**Comment by critchie** [comment]
Yeah discussion is good. I hope I’m not coming off as combative. I’ve been doing UI/UX work for over 20 years and just want to help out all I can.

Reading that document I think “pane” would have to be defined. I see the main screen of MS4 being divided into panes. The main editing pain, the tab pane, the palette pane, etc. Tabbing between those and then using the arrow keys feels natural here.

Reading all three topics in that link: Navigating in a Logical Order, Navigating Between Window Panes and Navigating Within Fields are the key areas for this discussion.

From the Navigating in a Logical Order:
“In Windows applications, users navigate by pressing the TAB key to move the input focus from one UI element to another. They press the SPACEBAR or ENTER key to choose the currently selected active region or to activate a control or command. Pressing the SHIFT+TAB key combination reverses the navigation order, moving the input focus backward through the elements, and pressing the arrow keys moves the input focus in specific directions within a group of elements.”

The last part about using arrows is a reference, for example, to a group of radio buttons. If you have four radio buttons in a radio group a TAB will put focus on the first one and then the arrow keys let you cycle through them. Or, like in MS4, a group of tabs like Home/Score/Publish/DevTools would use arrow keys for navigation.

But in the New Score dialog I don’t feel like there are panes here. The four items at the top: Key, Tempo, etc, to me, are just fancy drop downs and navigation between these should be handled with the Tab key. What is here is a dialog box with a bunch of UI elements.

I do like the way the Key signature component works. ENTER or SPACE opens it up with focus on the Major/Minor tabs. Use the arrow keys to pick Major/Minor. Then tab to the signature list which is just a pictorial group of radio buttons that uses arrows for navigation. SPACE selects the signature. ESC closes the “drop down”. Very nicely done.

But, in the Time signature there is an issue using the arrow keys for navigation. If I select the first signature, right arrow to the number scroller the focus/cursor is now captured. I can’t right or left arrow out of this field without changing the value. If I use the up or down arrow and change the value the right or left arrow will suddenly allow me to move to the next field. If the TAB key were used to move between these fields this would not be a problem. There is a group of radio buttons here so the up or down arrow key is allowed to navigate between the options.

The Tempo and Measures have similar issues where there is a mix of tabbing and arrow keys for navigation and number scroller fields that capture the focus/cursor.

**Comment by MarcSabatella** [comment]
@shoogle @bkunda what do you think?

I will add that during the demo I did for my MuseScore Café last week, got here and also thought it was simply broken when Tab didn't work.  Then eventually - like, after the stream was over - went to file a bug, and realized what was going on.  And as I played around with the interface more, it started to make sense, as it does seem fairly consistent.

And yet most likely lots of people will encounter that same bit of surprise pretty early on when they first try to create a score. Even assuming we continue to use this language elsewhere, it could be worth revisiting for this window specifically.

**Comment by shoogle** [comment]
Our hierarchical navigation system does bend the rules a bit when it comes to standard UI practice, but I think it's worth it for the massive improvement to navigation speed (once you get used to it) compared to a traditional flat/linear navigation system that uses the tab key to reach each individual control rather than just control groups.

In MuseScore 3 it took hundreds of presses of the tab key to get around the whole UI, but now in MuseScore 4 it's just 10 or so presses of the tab key to get around the control groups, and then you use the arrow keys to navigate within the groups, so it's much faster, at least in the main score window. Admittedly the new system isn't so crucial in the New Score dialog where there are far fewer controls than in the score window, but we need to maintain consistency within the application. It wouldn't make sense to navigate input fields one way in the Properties panel and a different way in the dialogs, for example. (The Format > Style dialog still uses the old navigation system because it has not yet been ported from C++ to QML, but when this happens it too will use the new navigation system.)

However, I do agree that there are some places where navigation is unintuitive or even a bit broken (see #11323 for example) so I think there is still room for tweaking in this area to get the best possible result. It might be that we need to revert to using the tab key to navigate away from input fields but still use the arrow keys for other controls like buttons and checkboxes, although I would like to avoid this if possible due to the large number of input fields in places like the Properties panel.

**Comment by critchie** [comment]
I think you are going against years and years and years of HMI tradition and one that is still recommended by the major OS makers. The user, via muscle memory, is always going to go for that TAB key. The team is taking it upon itself to completely redefine the Data Form metaphor which I don’t think is necessary.

Other solution or idea:
Back in the day controls were grouped using a GroupBox control that would provide a frame and a caption that could be assigned a keyboard mnemonic using a text value such as “&Score Composer”. The user then could press ALT+S to navigate directly to that group. The focus would be placed in the first control and TAB would take over from there.

**Comment by panthen** [comment]
Just to add, as a fresh voice, that I had a bug report written about this but thought I ought to double check it didn't already exist before posting.
Using the up-down arrow keys doesn't feel natural for this; ergonomically it is worse as the arrow keys are further away from the typing position meaning you have to move the hand to a new position before moving it back again. TAB is the habitual go-to for multi-box text input.
TL;DR: It _feels_ like a bug, not a consistency feature.

**Comment by bkunda** [comment]
I must admit this issue has also bothered me for a while too, and I do wonder whether breaking away from strict consistency in places like the input fields in the new score dialog might actually improve the UX for the majority of users – although of course I wouldn't want to do this if it would compromise the experience of users who rely on keyboard navigation and strictly consistent navigational practices. 

As it happens, I did just encounter a problematic experience with input fields in the properties panel that I think relates to the issue described in #11323, which is that the cursor gets stuck on input fields; in this case, I seem to be able to progress to the next input field _only_  if I change the value in the field (If I leave the value unchanged, I can't use the arrow keys to move away from the field). This situation did make me wonder whether using the tab key would not be more intuitive as a way of moving away from the input field.

In saying this, I haven't had terribly much to do with keyboard navigation and accessibility, so I'm aware that I'm probably skimming the surface of this issue. But my opinion on consistency is that it can sometimes make sense to break away from it if it serves the interests of the user and how they expect to be able to use the app.

**Comment by critchie** [comment]
#11323 was addressed and corrected by using tabs to navigate between the fields. So now there is an inconsistency when moving between different types of input components. Some will use arrows while another will use a tab.

**Comment by shoogle** [comment]
I think it is ok to use tab as a way to get out of a text field. It was in the design from the beginning that tab would be used when the arrow keys are not available (i.e. when they are needed for something else like text navigation), but perhaps we should choose to interpret this as when _any_ arrow key is not available rather than just when they are _all_ unavailable.

In the case of a single line text field, only the left and right keys are needed for cursor movement, so the up and down keys are still free for navigation. I see no harm in allowing the up key, down key, and the tab key to _all_ be used as ways to navigate away from the text field in this situation.

**Comment by bkunda** [comment]
> I think it is ok to use tab as a way to get out of a text field. It was in the design from the beginning that tab would be used when the arrow keys are not available (i.e. when they are needed for something else like text navigation), but perhaps we should choose to interpret this as when _any_ arrow key is not available rather than just when they are _all_ unavailable.
> 
> In the case of a single line text field, only the left and right keys are needed for cursor movement, so the up and down keys are still free for navigation. I see no harm in allowing the up key, down key, and the tab key to _all_ be used as ways to navigate away from the text field in this situation.

Thanks @shoogle. Can you also please amend the issue description so our devs know exactly how to fix this?
Many thanks!

**Comment by ecstrema** [comment]
Lots of work has definitely been put into accessibility for Mu4, and it is great to see such a thing being tackled.

However, implementing such accessibility from scratch has made room for weird/inconsistent behaviour; IMHO, too many behaviours have been added that negatively impact the regular user. This arrow-up-down between text fields is in my opinion one of these. I haven't see that behaviour anywhere else, and if it wasn't for the fact that I _knew_ there was a way to navigate between the fields with the keyboard, I wouldn't have found it. It definitely is not what any user would expect.

It may be because my focus has been a lot on web development lately, where tab and alt are sufficient to do _almost_ everything, but, as @critchie pointed out, it feels like it has deviated from OS-recommended practices.

**Comment by frfancha** [comment]
How is it possible to use arrow to move from field to field when some of these fields are drop down list requiring arrow to navigate between their values ?
Only tab can make sense.
Perhaps in that specific dialog box there are no drop down, only text fields, but that's no reason to break the rule.
Any form having some radio buttons or drop down must keep arrows dedicated to navigation in drop down and radio buttons values, and therefore must use tab to go from field to field.
That's not to say that providing some key for high level navigation allowing to quickly exit forms with lot of fields or to navigate through the whole UI of MuseScore in 10 key presses like Shoogle said is a bad idea.
Surely not, it is a great Idea!
But using tab for that is a bad idea.

**Comment by MarcSabatella** [comment]
Spin boxes and dropdowns do generally want tab indeed, and that seems to work in the places I tried.  if you find a specific control that isn't working, definitely open a separate issue for that.

And to be clear, Tab isn't the only key in the hierarchy, there is also F6 to move in larger groups still (separate panes/subwindows).

**Comment by frfancha** [comment]
I don't think you understood what I mean.
Probably I have explained incorrectly, let's rephrase:
Yes in form having spin boxes and drop downs Musescore had to keep tab as navigation key from field to field, I know that.
But that's why in form not having such fields, as is the case here, tab should be used as well. No reason that the user must guess "oh there is no spin box here so let's use arrow"

**Comment by MarcSabatella** [comment]
In theory there shouldn't be that kind of guesswork required; the nature of the field itself should tell you.  But again, if there is a specific item you have a concern about, best to open a separate issue for that specific element.

**Comment by frfancha** [comment]
This is not a separate topic, it is exactly this topic

**Comment by MarcSabatella** [comment]
This topic is about a specific dialog.  again, if there is a specific control in this specific dialog you have an issue with, please explain.  Otherwise, *please* continue the discussion elsewhere.  It is not productive to derail this discussion of this *specific* further than it already is.

**Comment by marczellm** [comment]
I use keyboard navigation in basically all apps all the time. This dialog in MS4 definitely feels broken because of this. No other app has required me to use arrow keys to move between text boxes, and I had no discoverable way to find out that this one does.

I think that one aspect worth considering is how "custom" a given UI feels to the user. If it feels like something very standard, such as a row of text boxes in an everyday looking dialog, then it should be navigated the standard way. As we move up the "customness" axis in terms of how the UI looks and acts different from usual UIs, it becomes more and more acceptable for the user to have to learn custom navigation keys.

**Comment by jrbowden** [comment]
As @critchie  and others have said, using Arrows to move from control to control **within a dialog box** is highly non-standard and does cause user confusion. I know this particular issue is about a particular control in a particular dialog, but imho:
The F6 / Tab / Arrows, work well for the main window of MuseScore where there are potentially dozens and dozens of controls to move through. Well done folks for doing this.
But, this should definitely not be the way to navigate through dialog boxes. It is very non-standard and will leave people scratching their heads how to accomplish even basic tasks.
e.g. Open the New Score Wizard and Tab around the controls, and you only reach the Cancel button: you can't move beyond the first screen.
At the other end, quitting MuseScore, Tabbing and you only get to the Don't Save button.
I agree all dialogs need a more intuitive, regular use of Tab key.
Apologies for butting in and I hope I haven't caused any further confusion.

**Comment by iainhallam** [comment]
Came here to file this bug, but found it in searching. Hierarchical keyboard navigation for a main window is innovative, but tab is well-established to mean, "move to the next UI element", and I thought the New Score dialog was broken until I read the discussion here. I'd never have thought to use up and down - I even tried left and right arrows. I agree with @jrbowden in that, when navigating a simpler dialog, tab keys should be the norm.

**Comment by skdonnelly** [comment]
This is still an issue. Tabbing to toolbars and using arrow keys to go through the buttons is one thing, but a dialog with form fields is different. Currently when you open the Project properties dialog you cannot Tab to even the first field. Counter to any intuition, you can tab to the "Folder" button, bottom right, then you have to press the Down arrow key to jump your cursor UP to the Work title field. There is no visual grouping nor logical connection between the form fields and the "Folder" button, but they have been arbitrarily grouped, as have the OK/Cancel/New property buttons, so you can't tab to each individually. Worst case, the Tab key should include the "Work title" field in its tab order, but really there is no need on this dialog to do anything other than the standard thing of Tab going to each item in order, and arrow keys would be used for anything like dropdown lists, radio buttons etc., of which there are none.

Also regardless of arbitrary decision to use this "grouping" method, it is broken on this dialog. If you figure out how to get to the "Work title" field, then use down-arrow to go through the fields (I have yet to encounter any other application that does this, but maybe they exist) then when you get to the "Copyright" field, you are stuck. Hitting down arrow does not work in this field. Tab inserts tabs in the text itself, down arrow does nothing, up arrow does nothing. I discovered the very non-intuitive "Shift-Tab" will jump you to the OK button. But this is basically a keyboard trap, users should not have to try a bunch of things to figure out how to get out of the field.

Also because of this issue, it appears to be impossible to get to the Work number and subsequent fields using the keyboard. However, again through random experimentation, I figured out that if you tab to the "Folder" button, then press **left** arrow, you jump to the Audio.com URL, as opposed to the up arrow which would make some visual sense. Again users shouldn't have to try random things to figure out how to navigate.

More random things:
- when on the OK button, pressing right-arrow key cycles through OK/Cancel/New property repeatedly, and down arrow does the same thing. Left arrow does the opposite, but up-arrow stops at "New property" and doesn't cycle like the other 3 keys
- if you highlight New property and hit Spacebar, it adds a new row as expected. To get to this row, you have to hit Tab to highlight the Folder button, then left-arrow to get to the garbage can icon. Left-arrow goes to the "Value" field, but then Left-arrow no longer works. At this point you have to use Up-arrow instead to get to the Property field. You can then use the down-arrow key to get down to the FOlder button, but when you're there you can't use the up-arrow to get back to the fields you were just editing
- I think the New property button belongs at the bottom of the list of fields, rather than on its own at the bottom of the dialog, but that's just me. Could even just be a "+" button at the bottom of the list of fields
- I think the folder button should have a tooltip or description, I clicked it then couldn't find an "OK" button, because I thought this button was to change the folder that my project was saved in. Realized eventually it's just to open the folder that the project is in.

Attempts to "improve" the UI marginally have just resulted in creating a couple of absolute accessibility fails. If you use the New property button, there is also no way to get to the property field you've just created, because of the same issue. 

Similar issue with the Save dialog, there are only 3 UI elements, two buttons and a checkbox, and they are not visually grouped, so tabbing between the 3 should work. Constantly having to figure out when to use Tab and when to use arrow keys just messes with flow and causes confusion. Clearly grouped lists of items or buttons where you have to select one: use arrow keys. Everything else: Tab. 

Barring usability testing or user feedback that indicates that this non-standard paradigm is helpful, you should avoid "UX theatre" of pretending you know what users want and changing industry-standard practices. The referenced post from Joanie Diggs is specifically referring to toolbars, so expanding it to other UI elements doesn't automatically make sense. 

Happy to help with testing fixes if there is a way to contribute.

**Comment by jmwmusic** [comment]
Agree that in dialog boxes tab should always jump to the next actionable item or for radio buttons the default already selected choice for that group of options.  It should not skip items.  I have been using Musescore for years and still got caught out again recently in the Style Dialog not realising a whole bunch of options were there if you arrowed down.  This ticket is just one instance of #20056
